snapshots/overlay/overlay.go

```diff
@@ -102,22 +130,92 @@ func NewSnapshotter(root string, opts ...Opt) (snapshots.Snapshotter, error) {
 		return nil, err
 	}
 
-	if err := os.Mkdir(filepath.Join(root, "snapshots"), 0700); err != nil && !os.IsExist(err) {
-		return nil, err
+	firstSnapshot := true
+	if err := os.Mkdir(filepath.Join(root, "snapshots"), 0700); err != nil {
+		if !os.IsExist(err) {
+			return nil, err
+		}
+		firstSnapshot = false
 	}
 	// figure out whether "userxattr" option is recognized by the kernel && needed
 	userxattr, err := overlayutils.NeedsUserXAttr(root)
 	if err != nil {
 		logrus.WithError(err).Warnf("cannot detect whether \"userxattr\" option needs to be used, assuming to be %v", userxattr)
 	}
 
+	var (
+		metacopyOn   bool
+		metacopyCtrl bool
+	)
+	metacopyCtrl = config.metacopyCtrl && supportsMetacopy()
+	metacopyCacheState := cacheSupportsMetacopy(root)
+	if metacopyCacheState != metacopyCacheNone {
```

Comment:
Perhaps it would be clearer to use switch, eg.
```go
        switch cacheSupportsMetacopy(root) {
        case metacopyCacheTrue:
                // since there already are snapshots with metacopy=on
                // it is not allowed to create a new one with metacopy=off
                metacopyOn = true
                if userxattr {
                        logrus.Warnf("userxattr and metacopy=on can't be used together. UserXAttr disabled.")
                        userxattr = false
                }
        case metacopyCacheFalse:
                metacopyOn = false
        case metacopyCacheNone:
                ...
        default:
                return nil, fmt.Errorf("...some error...")
        }
```

Reply:
Thanks, will fix it